pkg/controller/lambda/function/setup.go

```diff
@@ -0,0 +1,499 @@
+		!reflect.ValueOf(cr.Spec.ForProvider.Environment.Variables).IsZero() {
+		envVars = cr.Spec.ForProvider.Environment.Variables
+	}
+	if !reflect.ValueOf(obj.Configuration.Environment).IsZero() &&
+		!reflect.ValueOf(obj.Configuration.Environment.Variables).IsZero() {
+		awsVars = obj.Configuration.Environment.Variables
+	}
+	// Check if counts match which should handle nil/empty
+	if len(envVars) != len(awsVars) {
+		return false
+	}
+	// Check if actual values match, and only DeepEqual if there are actual values
+	if len(envVars) > 0 &&
+		!reflect.DeepEqual(envVars, awsVars) {
+		return false
+	}
+	return true
+}
+
+func isUpToDateFileSystemConfigs(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// Handle nil pointer refs
+	var fileSystemConfigs []*svcsdk.FileSystemConfig
+	var awsFileSystemConfigs []*svcsdk.FileSystemConfig
+	if !reflect.ValueOf(cr.Spec.ForProvider.FileSystemConfigs).IsZero() {
+		for _, v := range cr.Spec.ForProvider.FileSystemConfigs {
+			fileSystemConfigs = append(fileSystemConfigs, &svcsdk.FileSystemConfig{
+				Arn:            v.ARN,
+				LocalMountPath: v.LocalMountPath,
+			})
+		}
+	}
+	if !reflect.ValueOf(obj.Configuration.FileSystemConfigs).IsZero() {
+		awsFileSystemConfigs = obj.Configuration.FileSystemConfigs
+	}
+	// Check if counts match which should handle nil/empty
+	if len(fileSystemConfigs) != len(awsFileSystemConfigs) {
+		return false
+	}
+	// Check if actual values match, and only DeepEqual if there are actual values
+	if len(fileSystemConfigs) > 0 &&
+		!reflect.DeepEqual(fileSystemConfigs, awsFileSystemConfigs) {
+		return false
+	}
+	return true
+}
+
+func isUpToDateCode(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// GetFunctionOutput returns
+	// Code *FunctionCodeLocation `type:"structure"`
+	// which does not map to
+	// Code *FunctionCode `type:"structure" required:"true"`
+	// which is used when creating the function.
+	// We can't currently properly implement a comparison
+	return true
+}
+
+func isUpToDateLayers(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// The generator is creating layers with type of
+	// 	Layers []*string `json:"layers,omitempty"`
+	// Instead of []*svcsdk.Layer
+	// We can't properly implement a comparison until that is fixed
+	return true
+}
+
+func isUpToDateTracingConfig(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// Handle nil pointer refs
+	mode := svcsdk.TracingModePassThrough
+	if !reflect.ValueOf(cr.Spec.ForProvider.TracingConfig).IsZero() &&
+		!reflect.ValueOf(cr.Spec.ForProvider.TracingConfig.Mode).IsZero() {
+		mode = *cr.Spec.ForProvider.TracingConfig.Mode
+	}
+	if !reflect.DeepEqual(mode, *obj.Configuration.TracingConfig.Mode) {
+		return false
+	}
+	return true
+}
+
+func isUpToDateSecurityGroupIDs(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// Handle nil pointer refs
+	var securityGroupIDs []*string
```

Comment:
Same comment as line 186.

Reply:
Because of  `omitempty` - security groups will initialize as nil instead of an empty list if they aren't set.
The current behavior treats nil as empty list and will remove security groups in aws if they they aren't defined in crossplane.
Otherwise it would always end up returning false and be in a constant update loop.  Is there a better way to initialize nil to empty value?  Is there a better way to handle this?

Reply:
`len` returns `0` if the array is `nil`, so it's usually a safer approach to check for length of arrays/maps instead of a nil check since it covers both empty and nil cases.

Reply:
makes sense.  Since I had to do this several places I use the method you mention in `compareSliceStringStringPointer`